org.scala-refactoring.library/src/main/scala/scala/tools/refactoring/common/PimpedTrees.scala

```diff
@@ -882,22 +883,66 @@ trait PimpedTrees {
    * object.
    */
   object ModifierTree {
+    import Flags._
+
     def unapply(m: global.Modifiers): Option[List[ModifierTree]] = {
       val flagSorter = Ordering.by((pair: (Long, Position)) => flagPosition(pair._1))
       val sortedMods = m.positions.toList.sorted(flagSorter)
 
-      Some(sortedMods map {
+      Some(sortedMods flatMap {
         // hack to get rid of override modifiers
         // couldn't figure out how to remove a flag from the positions map (michael)
         case (Flags.OVERRIDE, _) if (m.flags & Flags.OVERRIDE) == 0 =>
-          ModifierTree(0)
+          Seq(ModifierTree(0))
         case (flag, pos) if pos.isRange =>
-          ModifierTree(flag) setPos (pos withEnd (pos.end + 1))
+          val fixedEnd = {
+            if (flag == PRIVATE) fixEndForScopedAccessModifier(pos)
+            else pos.end + 1
+          }
+
+          /*
+           *  Unfortunately package private and protected vals/defs (like "private[pgk] val test = 2") might not be properly represented
+           *  in 'm.positions' (see https://www.assembla.com/spaces/scala-ide/tickets/1002446#/activity/ticket:). We therefore
+           *  add the associated modifier trees "by-hand" if needed.
+           */
+          val missingModifierTree = {
+            if (m.privateWithin.nonEmpty && pos.end - pos.start < 3) {
+              val srcAtModifierEnd = SourceWithMarker(pos.source.content, pos.start - 1).moveMarker(commentsAndSpaces.backward)
+
+              val srcAtModifierStart = srcAtModifierEnd.moveMarker(
+                  (("private" | "protected") ~
+                  commentsAndSpaces ~ bracketsWithContents ~ commentsAndSpaces).backward)
+
+              Some(ModifierTree(extractAccessModifier(flag)).setPos(pos.withStart(srcAtModifierStart.marker + 1).withEnd(srcAtModifierEnd.marker + 1)))
```

Comment:
Could maybe made more readable by having a method in SourceWithMarker that returns a position instead of the int-marker?